compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -2138,32 +2144,70 @@ fn prefetch_mir(tcx: TyCtxt<'_>) {
 // will allow us to slice the metadata to the precise length that we just
 // generated regardless of trailing bytes that end up in it.
 
-#[derive(Encodable, Decodable)]
 pub struct EncodedMetadata {
-    raw_data: Vec<u8>,
+    mmap: Option<Mmap>,
+    // We need to carry MaybeTempDir to avoid deleting the temporary
+    // directory while accessing the Mmap.
+    _temp_dir: Option<MaybeTempDir>,
 }
 
 impl EncodedMetadata {
     #[inline]
-    pub fn new() -> EncodedMetadata {
-        EncodedMetadata { raw_data: Vec::new() }
+    pub fn from_path(path: PathBuf, temp_dir: Option<MaybeTempDir>) -> std::io::Result<Self> {
+        let file = std::fs::File::open(&path)?;
+        let file_metadata = file.metadata()?;
+        if file_metadata.len() == 0 {
+            return Ok(Self { mmap: None, _temp_dir: temp_dir });
+        }
+        let mmap = unsafe { Some(Mmap::map(file)?) };
+        Ok(Self { mmap, _temp_dir: temp_dir })
+    }
+
+    #[inline]
+    pub fn empty() -> Self {
+        Self { mmap: None, _temp_dir: None }
     }
 
     #[inline]
     pub fn raw_data(&self) -> &[u8] {
-        &self.raw_data
+        self.mmap.as_ref().map(|mmap| mmap.as_ref()).unwrap_or_default()
+    }
+}
+
+impl<S: Encoder> Encodable<S> for EncodedMetadata {
+    fn encode(&self, s: &mut S) -> Result<(), S::Error> {
+        let slice = self.raw_data();
+        slice.encode(s)
     }
 }
 
-pub fn encode_metadata(tcx: TyCtxt<'_>) -> EncodedMetadata {
+impl<D: Decoder> Decodable<D> for EncodedMetadata {
+    fn decode(d: &mut D) -> Self {
+        let temp_dir = TempFileBuilder::new().prefix("decoded").tempdir().unwrap();
+        let temp_dir = MaybeTempDir::new(temp_dir, false);
+        let filename = temp_dir.as_ref().join("decoded");
+        let file = std::fs::File::create(&filename).unwrap();
+        let mut file = std::io::BufWriter::new(file);
+
+        let len = d.read_usize();
+        for _ in 0..len {
+            file.write(&[d.read_u8()]).unwrap();
+        }
+        file.flush().unwrap();
+
+        Self::from_path(filename, Some(temp_dir)).unwrap()
+    }
+}
+
+pub fn encode_metadata(tcx: TyCtxt<'_>, path: impl AsRef<Path> + Send) {
```

Comment:
Why not a simple `&Path`?
```suggestion
pub fn encode_metadata(tcx: TyCtxt<'_>, path: &Path) {
```

Reply:
Fixed by https://github.com/rust-lang/rust/pull/96544/commits/899caff9e698c9c90f5b95baac07cfb91888aef1 .